### ISSUE TITLE
Bug: [no-unused-vars] False positive when returning an assignment

### ISSUE THREAD

**Issue by Mitsunee** [eslint/eslint#20353]
### Environment

Node version: v24.11.0
npm version: v11.6.3
Local ESLint version: v9.36.0 (Currently used)
Global ESLint version: Not found
Operating System: linux 6.17.8-200.fc42.x86_64

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
rules: {
  "no-unused-vars": "warn"
}
```
</details>

```js
function _foobar() {
  let foo;

  if (Math.random()) {
    return (foo = { bar: true }); // 'foo' is assigned a value but never used.  ([no-unused-vars](https://eslint.org/docs/rules/no-unused-vars))
  }

  return;
}
```

### What did you expect to happen?

No warning should be given

### What actually happened?

A warning is given despite the result of the assignment being used as the return value of the function in the same line

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBbXCJlcnJvclwiXSAqL1xuXG5leHBvcnQgZnVuY3Rpb24gX2Zvb2JhcigpIHtcbiAgbGV0IGZvbztcblxuICBpZiAoTWF0aC5yYW5kb20oKSkge1xuICAgIHJldHVybiAoZm9vID0geyBiYXI6IHRydWUgfSk7XG4gIH1cblxuICByZXR1cm47XG59XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

This might seem a bit silly in isolation, but was simplified from a larger function where I created a `result` variable at the top to return various different results that should all match a given type interface to make narrowing of the result easier outside of the function.

This issue was mostly copypaste from a downstream issue in typescript-eslint (https://github.com/typescript-eslint/typescript-eslint/issues/11782), but has been confirmed to correctly emulate an upstream bug in the base eslint rule `no-unused-vars` as well. I have been asked to report this here, so downstream can make a decision based on yours.

**Comment by DMartens** [comment]
I think this behavior is correct as the rule also reports for not accessing a value which was overridden ([playground](https://eslint.org/play/#eyJ0ZXh0IjoibGV0IGZvbyA9IHRydWU7XG5mb28gPSBmYWxzZSIsIm9wdGlvbnMiOnsicnVsZXMiOnsibm8tdW51c2VkLXZhcnMiOlsiZXJyb3IiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319LCJzb3VyY2VUeXBlIjoic2NyaXB0In19fQ==)). It does not matter that the assignment value is returned as the assignment to the variable in this case does nothing (meaning `return foo = VALUE` is always equivalent to `return VALUE` as the setting of the variable is not observable). Given your comment that the intention of assigning the value is to make sure the type matches a common interface, this makes more sense as a rule option in the extension rule of typescript-eslint where you initially reported it as I do not see a reason why the JavaScript rule should allow this.

**Comment by mdjermanovic** [comment]
This is not a bug in the core `no-unused-vars` rule. Assignment expression `left = right` always evaluates to `right` (the value of `left` is not read), so `return (foo = { bar: true });` doesn't actually read from `foo`, which means that the variable `foo` is never read in the original code example.

**Comment by mdjermanovic** [comment]
> This might seem a bit silly in isolation, but was simplified from a larger function where I created a `result` variable at the top to return various different results that should all match a given type interface to make narrowing of the result easier outside of the function.

As for the TypeScript-specific use case (repro code from https://github.com/typescript-eslint/typescript-eslint/issues/11782), we are in the process of implementing support for TypeScript syntax in core rules (https://github.com/eslint/eslint/issues/19173; `no-unused-vars` is included), so an exception or an option for this use case would eventually end up being implemented in the core `no-unused-vars` rule.

Is it common to declare variables just for type checking? Couldn't the same be achieved this way:

```ts
function _foobar() {
  type foo = { bar: boolean };

  if (Math.random()) {
    return { bar: true } satisfies foo;
  }

  return;
}
```

**Comment by Mitsunee** [comment]
> Is it common to declare variables just for type checking? Couldn't the same be achieved this way:

I just played with this a bit more in typescript's playground and noticed that both my solution and `satisfies` have the *same* issue where the inferred return is the `typeof` of the returned value, losing the interface in the process, so I guess that makes me suggestion even more silly.

I think this comes down to a misunderstanding of mine of how the `return (foo = value)` syntax works. The variable is merely assigned and the assigned value is *also* `return`ed, leaving the variable not used, only assigned, so actually this isn't a false positive at all and also doesn't achieve the TS behaviour I wanted, so here it *would* be better to use an explicit return type, which I try to avoid as it can lead to "lying" to TypeScript.

I'll close my issues, thanks for the feedback and suggestion
